### PR TITLE
review panel test timer accesibility ACTP-305

### PR DESCRIPTION
**Pull request by lecosson**
https://oat-sa.atlassian.net/browse/ACTP-305

Allow a screen readers to access a timers of time-limited tests

**How to test:**
- switch taoAct in TAO instance to `feature/ACTP-305/timer-review-panel-aria` branch
- disable current key navigation plugin to acces native key navigation flow
- switch `oat-sa/tao-test-runner-qti-fe` in `taoQtiTest` extension to `feature/ACTP-305/timer-review-panel-aria` branch
- prepare a test with enabled timers
- login to ACT as a Guest or as user and run this test
- try to acces review panel from test using `TAB` key
- check a review panel is readed by your accessibility software, including timers value

**Related pull request**
https://github.com/oat-sa/extension-tao-act/pull/1340

### CODE REVIEW

File: src/plugins/controls/timer/component/countdown.js [hunk 1]
```diff
@@ -117,8 +120,10 @@ export default function countdownFactory($container, config) {
                         encodedTime = timeEncoder.encode(this.remainingTime / precision);
                         if (encodedTime !== this.encodedTime) {
                             this.encodedTime = encodedTime;
+                            [hours, minutes/* , seconds */] = this.encodedTime.split(':');
```
Comment (jsconan) [inline src/plugins/controls/timer/component/countdown.js]: Is it intended to keep dead code? You should remove it.
Why mixing ES6 and ES5?

File: src/plugins/controls/timer/component/countdown.js [hunk 2]
```diff
@@ -273,9 +278,13 @@ export default function countdownFactory($container, config) {
         })
         .on('render', function() {
             $time = $('.time', this.getElement());
+            $timeAriaLabel = $('#timerbox-aria-label', this.getElement());
+            const timeString = timeEncoder.encode(this.remainingTime / precision);
+            const [hours, minutes/* , seconds */] = timeString.split(':');
```
Comment (jsconan) [inline src/plugins/controls/timer/component/countdown.js]: dead code

File: package.json [hunk 3]
```diff
@@ -12,11 +12,11 @@
         "test": "npx qunit-testrunner",
         "test:keepAlive": "npx qunit-testrunner --keepalive",
         "test:cov": "npm run build:cov && npx qunit-testrunner --cov",
-        "test:dev": "NODE_ENV=development run-p test:keepAlive build:watch",
+        "test:dev": "cross-env NODE_ENV=development run-p test:keepAlive build:watch",
```
Comment (jsconan) [inline package.json]: Is this addition related to the feature? If no, you should rather add it in a separate PR

File: src/plugins/controls/timer/component/countdown.js [hunk 4]
```diff
@@ -117,8 +119,12 @@ export default function countdownFactory($container, config) {
                         encodedTime = timeEncoder.encode(this.remainingTime / precision);
                         if (encodedTime !== this.encodedTime) {
                             this.encodedTime = encodedTime;
+                            const timeArr = this.encodedTime.split(':');
+                            const hours = timeArr[0];
+                            const minutes = timeArr[1];
 
                             $time.text(this.encodedTime);
+                            $timeAriaLabel.text( __('timer: %s hours %s minutes to answer', hours, minutes ));
```
Comment (ampaveliev) [inline src/plugins/controls/timer/component/countdown.js]: dont see when the `$timeAriaLabel` variable was initialized

File: src/plugins/controls/timer/component/scss/timerbox.scss [hunk 5]
```diff
@@ -8,6 +8,10 @@
         @include transition(opacity, 300ms, linear);
     }
 
+    & #timerbox-aria-label {
+        display:none;
```
Comment (ampaveliev) [inline src/plugins/controls/timer/component/scss/timerbox.scss]: Space after colon. Also display none make element invisible for the screen readers, does it make a sense?

File: src/plugins/controls/timer/component/countdown.js (hunk 4 above)
Comment (lecosson) [reply src/plugins/controls/timer/component/countdown.js]: declared in begin (line 92), initialized on render (line 282)

File: src/plugins/controls/timer/component/scss/timerbox.scss (hunk 5 above)
Comment (lecosson) [reply src/plugins/controls/timer/component/scss/timerbox.scss]: used to provide data in another plugin via `aria-labelledby`

File: src/plugins/controls/timer/component/countdown.js [hunk 6]
```diff
@@ -273,9 +279,15 @@ export default function countdownFactory($container, config) {
         })
         .on('render', function() {
             $time = $('.time', this.getElement());
+            $timeAriaLabel = $('#timerbox-aria-label', this.getElement());
+            const timeString = timeEncoder.encode(this.remainingTime / precision);
+            const timeArr = this.encodedTime.split(':');
```
Comment (jsconan) [inline src/plugins/controls/timer/component/countdown.js]: This is generating a runtime error: 
```javascript
TypeError: this.encodedTime is undefined' from http://act.tao/taoQtiTest/views/node_modules/@oat-sa/tao-test-runner-qti/dist/plugins/controls/timer/component/countdown.js?buster=5e73391bc0bd3:301:23 Error: "Caught[via window.onerror]: 'TypeError: this.encodedTime is undefined' from http://act.tao/taoQtiTest/views/node_modules/@oat-sa/tao-test-runner-qti/dist/plugins/controls/timer/component/countdown.js?buster=5e73391bc0bd3:301:23
```

Comment (lecosson) [reply src/plugins/controls/timer/component/countdown.js]: updated, added checking for missed properties

File: src/plugins/controls/timer/component/countdown.js [hunk 7]
```diff
@@ -272,10 +282,21 @@ export default function countdownFactory($container, config) {
...
             $time = $('.time', this.getElement());
+            $timeAriaLabel = $('#timerbox-aria-label', this.getElement());
 
             if (this.config.showBeforeStart === true) {
-                $time.text(timeEncoder.encode(this.remainingTime / precision));
+                if (timeString) {
+                    timeArr = timeString.split(':');
+                    hours = timeArr[0];
+                    minutes = timeArr[1];
+                    $timeAriaLabel.text( __('timer: %s hours %s minutes to answer', hours, minutes ));
```
Comment (jsconan) [inline src/plugins/controls/timer/component/countdown.js]: Since the same code is presented twice, please consider factoring it in a single place.

File: src/plugins/controls/timer/component/countdown.js [hunk 8]
```diff
@@ -272,10 +282,21 @@ export default function countdownFactory($container, config) {
             this.render($container);
         })
         .on('render', function() {
+            var timeString = timeEncoder.encode(this.remainingTime / precision);
+            var timeArr;
+            var hours;
+            var minutes;
             $time = $('.time', this.getElement());
+            $timeAriaLabel = $('#timerbox-aria-label', this.getElement());
```
Comment (jsconan) [inline src/plugins/controls/timer/component/countdown.js]: It should be better to rely on a class name to access the element.

File: src/plugins/controls/timer/component/scss/timerbox.scss [hunk 9]
```diff
@@ -8,6 +8,10 @@
         @include transition(opacity, 300ms, linear);
     }
 
+    & #timerbox-aria-label {
+        display: none;
```
Comment (jsconan) [inline src/plugins/controls/timer/component/scss/timerbox.scss]: This prevents the screen reader to announce the content (at least under Firefox + NVDA).
Since the element is nested inside another one, no need to use the parent alias `&`.
It should also be better to use a class instead of an ID in CSS, even if the aria attribute relies on the ID.

File: src/plugins/controls/timer/component/tpl/countdown.tpl [hunk 10]
```diff
@@ -1,4 +1,5 @@
 <div class="countdown" data-control="{{id}}" data-type="{{type}}" data-scope="{{scope}}" title="{{label}}" disabled>
     <span class="label truncate">{{label}}</span>
     <span class="time"></span>
+    <span id="timerbox-aria-label"></span>
```
Comment (jsconan) [inline src/plugins/controls/timer/component/tpl/countdown.tpl]: Unfortunately the ARIA attributes only work with ID. However, it should be prefixed to make sure the ID will only relate to the current test runner.

File: src/plugins/controls/timer/component/timerbox.js [hunk 11]
```diff
@@ -212,7 +213,12 @@ export default function timerboxFactory(config) {
                                 self.trigger('timerend', self.timers[id]);
                             })
                             .on('change', function(value) {
+                                var precision, encodedTime;
                                 if (self.timers[id]) {
+                                    precision = 1000; // precision is milliseconds
+                                    encodedTime = timeEncoder.encode(this.remainingTime / precision).split(':');
```
[the remark this answers is not in the document]
Comment (lecosson) [reply src/plugins/controls/timer/component/timerbox.js:219]: yes, it is reasonable, refactored

File: src/plugins/controls/timer/plugin.js [hunk 12]
```diff
@@ -215,6 +215,9 @@ export default pluginFactory({
                             .on('init', resolve)
                             .on('error', handleError);
 
+                        // share this timer values to use in other components
+                        self.timerbox.spread(testRunner, ['timertick']);
```
[the remark this answers is not in the document]
Comment (lecosson) [reply src/plugins/controls/timer/plugin.js:219]: fixed
